server/etcdserver/api/etcdhttp/readiness.go

```diff
@@ -0,0 +1,211 @@
+// Copyright 2023 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcdhttp
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"go.uber.org/zap"
+
+	"go.etcd.io/etcd/api/v3/etcdserverpb"
+	"go.etcd.io/etcd/server/v3/auth"
+)
+
+func InstallHealthEndpoints(lg *zap.Logger, mux *http.ServeMux, server ServerHealth) {
+	readyz := ReadyzCheckRegistry(server)
+	readyz.InstallHttpEndpoints(mux, lg)
+	livez := LivezCheckRegistry(server)
+	livez.InstallHttpEndpoints(mux, lg)
+}
+
+type HealthCheck func(ctx context.Context) error
+
+type CheckRegistry struct {
+	path string
+
+	checks map[string]HealthCheck
+}
+
+func LivezCheckRegistry(server ServerHealth) *CheckRegistry {
+	reg := CheckRegistry{path: "/livez", checks: make(map[string]HealthCheck)}
+	reg.Register("ping", ping)
+	reg.Register("serializable_read", checkSerializableRead(server))
+	return &reg
+}
+
+func ReadyzCheckRegistry(server ServerHealth) *CheckRegistry {
+	reg := CheckRegistry{path: "/readyz", checks: make(map[string]HealthCheck)}
+	reg.Register("ping", ping)
+	reg.Register("data_corruption", checkAlarm(server, etcdserverpb.AlarmType_CORRUPT))
+	reg.Register("serializable_read", checkSerializableRead(server))
+	return &reg
+}
+
+func (reg *CheckRegistry) Register(name string, check HealthCheck) {
+	reg.checks[name] = check
+}
+
+func (reg *CheckRegistry) InstallHttpEndpoints(mux *http.ServeMux, lg *zap.Logger) {
+	lg.Info(fmt.Sprintf("Installing %s http endpoint.", reg.path))
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	cancel()
+	mux.Handle(reg.path, reg.handleRootHealth(ctx, lg))
+	for checkName, check := range reg.checks {
+		mux.Handle(fmt.Sprintf("%s/%v", reg.path, checkName), adaptCheckToHandler(ctx, check))
+	}
```

Comment:
It does not look correct. I think the health checker needs to respect client request context, instead of the hard-coded timeout configured by server. 

Please check  https://github.com/etcd-io/etcd/pull/16651#discussion_r1340716818 and https://github.com/etcd-io/etcd/pull/16651#discussion_r1340718539 and https://pkg.go.dev/net/http#Request.Context

Could you please write a integration test to validate it? Let's say a client disconnects right after sending the requests to server, will server waste the time to do health checks.